lib/actions-oclif/preload.ts

```diff
@@ -0,0 +1,549 @@
+	];
+
+	public static args = [
+		{
+			name: 'image',
+			description: 'the image file path',
+			required: true,
+		},
+	];
+
+	public static usage = 'preload <image>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		app: flags.string({
+			description: 'name, slug or numeric ID of the application to preload',
+			char: 'a',
+		}),
+		commit: flags.string({
+			description: `\
+The commit hash for a specific application release to preload, use "current" to specify the current
+release (ignored if no appId is given). The current release is usually also the latest, but can be
+manually pinned using https://github.com/balena-io-projects/staged-releases .\
+`,
+			char: 'c',
+		}),
+		'splash-image': flags.string({
+			description: 'path to a png image to replace the splash screen',
+			char: 's',
+		}),
+		'dont-check-arch': flags.boolean({
+			description:
+				'disables check for matching architecture in image and application',
+		}),
+		'pin-device-to-release': flags.boolean({
+			description:
+				'pin the preloaded device to the preloaded release on provision',
+			char: 'p',
+		}),
+		'add-certificate': flags.string({
+			description: `\
+Add the given certificate (in PEM format) to /etc/ssl/certs in the preloading container.
+The file name must end with '.crt' and must not be already contained in the preloader's
+/etc/ssl/certs folder.
+Can be repeated to add multiple certificates.\
+`,
+		}),
+		...dockerConnectionCliFlags,
+		// Redefining --dockerPort here (defined already in dockerConnectionCliFlags)
+		// without -p alias, to avoid clash with -p alias of pin-device-to-release
+		dockerPort: flags.integer({
+			description:
+				'Docker daemon TCP port number (hint: 2375 for balena devices)',
+			parse: (p) => parseAsInteger(p, 'dockerPort'),
+		}),
+		// Not supporting -h for help, because of clash with -h in DockerCliFlags
+		// Revisit this in future release.
+		help: flags.help({}),
+	};
+
+	public static authenticated = true;
+
+	public static primary = true;
+
+	public async run() {
+		const { args: params, flags: options } = this.parse<FlagsDef, ArgsDef>(
+			PreloadCmd,
+		);
+
+		const balena = getBalenaSdk();
+		const balenaPreload = await import('balena-preload');
+		const visuals = getVisuals();
+		const nodeCleanup = await import('node-cleanup');
+		const { ExpectedError, instanceOf } = require('../errors');
+
+		// Check image file exists
+		try {
+			const fs = await import('fs');
+			await fs.promises.access(params.image);
+		} catch (error) {
+			throw new ExpectedError(
+				`The provided image path does not exist: ${params.image}`,
+			);
+		}
+
+		const progressBars: any = {};
+
+		const progressHandler = function (event: any) {
+			let progressBar = progressBars[event.name];
+			if (!progressBar) {
+				progressBar = progressBars[event.name] = new visuals.Progress(
+					event.name,
+				);
+			}
+			return progressBar.update({ percentage: event.percentage });
+		};
+
+		const spinners: any = {};
```

Comment:
```suggestion
		const spinners: {[key: string]: visuals.Spinner}= {};
```

Reply:
How should that be done when `visuals = getVisuals()`?  Do we need a seperate `import type`?

Reply:
Yeah it should be possibly to do this via an `import type`, worst case it can be done with something along the lines of:
```suggestion
		const spinners: {[key: string]: ReturnType<typeof getVisuals>['Spinner']}= {};
```